File: web/src/api/tokens.ts
import { request } from './client'
import type { components } from './schema'

export type Token = components['schemas']['Token']
export type TokenList = components['schemas']['TokenList']
export type TokenCreateResponse = components['schemas']['TokenCreateResponse']

export function listTokens(): Promise<TokenList> {
  return request<TokenList>('/api/v1/tokens')
}

export function getToken(id: number | string): Promise<Token> {
  return request<Token>(`/api/v1/tokens/${id}`)
}

export interface CreateTokenParams {
  name: string
  role: string
  expiresAt?: string
}

// Returns the plaintext token - shown exactly once, never retrievable again.
export function createToken(params: CreateTokenParams): Promise<TokenCreateResponse> {
  return request<TokenCreateResponse>('/api/v1/tokens', { method: 'POST', body: params })
}

export function revokeToken(id: number | string): Promise<void> {
  return request<void>(`/api/v1/tokens/${id}`, { method: 'DELETE' })
}

File: internal/client/tokens.go
package client

import (
	"context"
	"fmt"
	"net/http"
	"time"
)

// Token mirrors the Token schema (task 8.3/8.7).
type Token struct {
	ID        int64      `json:"id"`
	Name      string     `json:"name"`
	Role      string     `json:"role"`
	TokenHint string     `json:"tokenHint"`
	CreatedAt string     `json:"createdAt"`
	ExpiresAt *time.Time `json:"expiresAt,omitempty"`
	RevokedAt *string    `json:"revokedAt,omitempty"`
}

type TokenList struct {
	Items []Token `json:"items"`
}

type CreateTokenParams struct {
	Name      string     `json:"name"`
	Role      string     `json:"role"`
	ExpiresAt *time.Time `json:"expiresAt,omitempty"`
}

// CreateTokenResult carries the plaintext token exactly once - not
// retrievable again after this response.
type CreateTokenResult struct {
	Token
	PlaintextToken string `json:"token"`
}

func (c *Client) ListTokens(ctx context.Context) (TokenList, error) {
	var list TokenList
	err := c.do(ctx, http.MethodGet, "/api/v1/tokens", requestOptions{}, &list)
	return list, err
}

func (c *Client) GetToken(ctx context.Context, id int64) (Token, error) {
	var tok Token
	err := c.do(ctx, http.MethodGet, fmt.Sprintf("/api/v1/tokens/%d", id), requestOptions{}, &tok)
	return tok, err
}

func (c *Client) CreateToken(ctx context.Context, params CreateTokenParams) (CreateTokenResult, error) {
	var result CreateTokenResult
	err := c.do(ctx, http.MethodPost, "/api/v1/tokens", requestOptions{body: params}, &result)
	return result, err
}

// RevokeToken soft-revokes a token principal - never a hard delete, same
// reasoning as RevokeUser.
func (c *Client) RevokeToken(ctx context.Context, id int64) error {
	return c.do(ctx, http.MethodDelete, fmt.Sprintf("/api/v1/tokens/%d", id), requestOptions{}, nil)
}
